### DISCUSSION TITLE
Failed to Display a clinical image on the canvas in order to overlay annoatations and drawings on it

### DISCUSSION THREAD

**Discussion by jona42-ui** [fabricjs/fabric.js#9378] [Q&A]
hello
I am implementing an annotation feature(am making a custom svg editor using fabric), but using the `canvas.add(image)`  does not seem to render the image on the canvas. kindly i need help
<details> 
`import React, { useEffect, useRef, useState } from "react";
import { fabric } from "fabric";

const SvgEditor = () => {
  const canvasRef = useRef(null);
  const [canvas, setCanvas] = useState(null);
  const [drawingMode, setDrawingMode] = useState("rectangle");
  const [stateHistory, setStateHistory] = useState([]);
  const [currentStatePointer, setCurrentStatePointer] = useState(-1);
  const [isDrawing, setIsDrawing] = useState(false);
  const [lastPosition, setLastPosition] = useState({ x: 0, y: 0 });

  useEffect(() => {
    const options = {
      width: 800,
      height: 600,
    };
    const newCanvas = new fabric.Canvas(canvasRef.current, options);
    setCanvas(newCanvas);
    newCanvas.isDrawingMode = drawingMode === "freehand";

    newCanvas.on("mouse:down", handleMouseDown);
    newCanvas.on("mouse:move", handleMouseMove);
    newCanvas.on("mouse:up", handleMouseUp);

    saveCanvasState();

    return () => {
      newCanvas.dispose();
    };
  }, [drawingMode]);

  const selectDrawingMode = (mode) => {
    setDrawingMode(mode);
    canvas.isDrawingMode = mode === "freehand";
  };

  const addShape = () => {
    let shape;
    if (drawingMode === "rectangle") {
      shape = new fabric.Rect({
        width: 100,
        height: 50,
        fill: "transparent",
        stroke: "blue",
        strokeWidth: 2,
        left: 100,
        top: 100,
      });
    } else if (drawingMode === "circle") {
      shape = new fabric.Circle({
        radius: 25,
        fill: "transparent",
        stroke: "red",
        strokeWidth: 2,
        left: 200,
        top: 200,
      });
    }
    if (shape) {
      canvas.add(shape);
      saveCanvasState();
    }
  };

  const addText = () => {
    const text = new fabric.IText("Type your text here", {
      left: 300,
      top: 300,
      fill: "black",
    });
    canvas.add(text);
    saveCanvasState();
  };

  const changeColor = (color) => {
    const activeObject = canvas.getActiveObject();
    if (activeObject) {
      activeObject.set({ fill: color });
      canvas.renderAll();
      saveCanvasState();
    }
  };

  const handleMouseDown = (event) => {
    if (drawingMode === "freehand") {
      setIsDrawing(true);
      const { offsetX, offsetY } = event.e;
      setLastPosition({ x: offsetX, y: offsetY });
    }
  };

  const handleMouseMove = (event) => {
    if (isDrawing) {
      const { offsetX, offsetY } = event.e;
      const path = new fabric.Path(
        `M ${lastPosition.x} ${lastPosition.y} L ${offsetX} ${offsetY}`,
        {
          stroke: "blue",
          strokeWidth: 2,
          fill: "transparent",
        }
      );
      canvas.add(path);
      setLastPosition({ x: offsetX, y: offsetY });
    }
  };

  const handleMouseUp = () => {
    setIsDrawing(false);
    saveCanvasState();
  };

  const saveCanvasState = () => {
    const canvasState = JSON.stringify(canvas);
    const newHistory = [
      ...stateHistory.slice(0, currentStatePointer + 1),
      canvasState,
    ];
    setStateHistory(newHistory);
    setCurrentStatePointer(newHistory.length - 1);
  };

  const undo = () => {
    if (currentStatePointer > 0) {
      const newPointer = currentStatePointer - 1;
      const canvasState = stateHistory[newPointer];
      setCurrentStatePointer(newPointer);
      canvas.loadFromJSON(canvasState, () => {
        canvas.renderAll();
      });
    }
  };

  const redo = () => {
    if (currentStatePointer < stateHistory.length - 1) {
      const newPointer = currentStatePointer + 1;
      const canvasState = stateHistory[newPointer];
      setCurrentStatePointer(newPointer);
      canvas.loadFromJSON(canvasState, () => {
        canvas.renderAll();
      });
    }
  };

  const handleImageUpload = (event) => {
    const file = event.target.files[0];
    if (file) {
      const reader = new FileReader();
      reader.onload = (e) => {
        // eslint-disable-next-line no-console
        console.log("Image loaded successfully!");
        const image = new fabric.Image();
        image.setElement(e.target.result);
        canvas.add(image);
        // eslint-disable-next-line no-console
        console.log(canvas.add(image));
        canvas.renderAll();
        saveCanvasState();
      };
      reader.readAsDataURL(file);
    }
  };

  return (
    <div>
      <canvas ref={canvasRef} width="800" height="600" />

      <button onClick={() => selectDrawingMode("rectangle")}>Rectangle</button>
      <button onClick={() => selectDrawingMode("circle")}>Circle</button>
      <button onClick={() => selectDrawingMode("freehand")}>Freehand</button>

      <button onClick={addShape}>Add Shape</button>
      <button onClick={addText}>Add Text</button>

      <input type="color" onChange={(e) => changeColor(e.target.value)} />

      <button onClick={undo}>Undo</button>
      <button onClick={redo}>Redo</button>

      <input type="file" accept="image/*" onChange={handleImageUpload} />
    </div>
  );
};

export default SvgEditor; <details>
`

**Comment by jona42-ui** [comment]
cc/ @ShaMan123

**Reply by ShaMan123** [reply]
If you want me to look then put more effort into this.
Make it into a codesandbox and remove these nasty dumps or at least use the details tag to make them collapse.
Until then....

**Reply by jona42-ui** [reply]
sorry about the noisy post

**Reply by asturur** [reply]
Yes noisy post are a bit of an issue.
Maybe we should link an example issue saying, `your issue should look like this one: link`.

The issue is for both you and us.
I go over the fabricJS email once a day when i can and when i get to your email, i can either read the issue and try to help or write you that the issue isn't good enough, done. The day is over for this issue and next time is tomorrow at best for a second check.

So you definitely lose a day
